File: python/148.py
"""Project Euler Problem 148.

We can easily verify that none of the entries in the first seven rows of Pascal's triangle are divisible by 7.

However, if we check the first one hundred rows, we will find that only 2361 of the 5050 entries are not divisible by 7.
Find the number of entries which are not divisible by 7 in the first one billion (10^9) rows of Pascal's triangle.
"""

BASE = 7


def to_base7(num: int) -> list:
    """Convert a number to its base-7 digits (most significant digit first)."""
    if num == 0:
        return []
    digits = []
    temp = num
    while temp > 0:
        digits.append(temp % BASE)
        temp //= BASE
    digits.reverse()
    return digits


class DigitDP:
    """Digit DP to count the total number of non-divisible entries in rows 0 to m."""

    def __init__(self, digits: list):
        """Initialize."""
        self.digits = digits
        self.memo = {}

    def dp(self, pos: int, tight: bool) -> int:
        """dp(pos, tight) returns the sum of products for all valid suffixes starting at position pos."""
        if pos == len(self.digits):  # Base case: empty suffix, product = 1
            return 1

        key = (pos, 1 if tight else 0)
        if key in self.memo:
            return self.memo[key]

        result = 0
        max_digit = self.digits[pos] if tight else (BASE - 1)

        for d in range(max_digit + 1):
            new_tight = tight and (d == max_digit)
            result += (d + 1) * self.dp(pos + 1, new_tight)

        self.memo[key] = result
        return result


def count_non_divisible(m: int) -> int:
    """Count non-divisible entries in rows 0 to m."""
    digits = to_base7(m)
    dp = DigitDP(digits)
    return dp.dp(0, True)


def main() -> int:
    """Main function."""
    # Count entries in rows 0 to n-1 (n rows total)
    # For first 10^9 rows, we need rows 0 to 10^9 - 1
    return count_non_divisible(10**9 - 1)


if __name__ == "__main__":
    print(main())

File: c/148.c
/*
 * Project Euler 148 - Entries in Pascal's triangle not divisible by 7
 *
 * Uses digit DP on base-7 representation.
 * For rows 0 to 10^9-1.
 */
#include <stdio.h>

int main(void) {
    long long N = 999999999LL; /* 10^9 - 1 */

    /* Convert to base 7 */
    int digits[20];
    int ndigits = 0;
    {
        long long tmp = N;
        while (tmp > 0) {
            digits[ndigits++] = (int)(tmp % 7);
            tmp /= 7;
        }
        /* Reverse */
        for (int i = 0; i < ndigits / 2; i++) {
            int t = digits[i];
            digits[i] = digits[ndigits - 1 - i];
            digits[ndigits - 1 - i] = t;
        }
    }

    /* Digit DP: dp(pos, tight) = sum of products of (digit+1) for all valid numbers */
    /* Iterative approach */
    /* result_tight = product considering tight constraint */
    /* result_free = accumulated from freed digits */

    long long result = 0;
    long long prefix_product = 1; /* product of (chosen_digit + 1) for tight path so far */

    for (int pos = 0; pos < ndigits; pos++) {
        int max_d = digits[pos];
        /* Sum contribution from digits 0..max_d-1 (which free the suffix) */
        /* For digit d (0 <= d < max_d), suffix is free: each position can be 0..6 */
        /* Free suffix product = 28^(remaining_positions) since sum_{d=0}^{6}(d+1) = 28 */
        long long free_suffix = 1;
        for (int j = pos + 1; j < ndigits; j++) free_suffix *= 28;

        long long sum_d = 0;
        for (int d = 0; d < max_d; d++) {
            sum_d += (d + 1);
        }
        result += prefix_product * sum_d * free_suffix;

        /* Continue tight path with digit max_d */
        prefix_product *= (max_d + 1);
    }
    /* Add the tight path itself (the number N exactly) */
    result += prefix_product;

    printf("%lld\n", result);
    return 0;
}
